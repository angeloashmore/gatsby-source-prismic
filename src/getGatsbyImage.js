// Default width for `fixed` images. Same as `gatsby-plugin-sharp`.
const DEFAULT_FIXED_WIDTH = 400

// Default resolutions for `fixed` images. Same as `gatsby-plugin-sharp`.
const DEFAULT_FIXED_RESOLUTIONS = [1, 1.5, 2]

// Default maxWidth for `fluid` images. Same as `gatsby-plugin-sharp`.
const DEFAULT_FLUID_MAX_WIDTH = 800

// Default breakpoint factors for `fluid` images. Same as
// `gatsby-plugin-sharp`.
const DEFAULT_FLUID_BREAKPOINT_FACTORS = [0.25, 0.5, 1.5, 2]

// Default quality for images. 50 is fairly aggressive.
const DEFAULT_QUALITY = 50

// Default image sizing strategy. `max` ensures the resulting image is never
// larger than the source file.
const DEFAULT_FIT = 'max'

const buildURL = (
  baseURL,
  {
    width,
    height,
    aspectRatio,
    rect,
    dpr,
    quality = DEFAULT_QUALITY,
    compress = true,
    format = true,
    fit = DEFAULT_FIT,
    blur,
  } = {},
) => {
  const params = {
    w: width,
    h: height,
    ar: aspectRatio,
    rect,
    fit,
    dpr,
    q: quality,
    blur,
    auto: encodeURIComponent(
      [compress && 'compress', format && 'format'].filter(Boolean).join(','),
    ),
  }

  const searchParams = Object.keys(params)
    .reduce((acc, key) => {
      const val = params[key]

      if (val === undefined || val === null || val.length < 1) return acc

      return [...acc, `${key}=${val}`]
    }, [])
    .join('&')

  return [baseURL, searchParams].join('?')
}

const buildBase64URL = (baseURL, params) =>
  buildURL(baseURL, { width: 100, blur: 15, quality: 20, ...params })

const buildFixedSrcSet = (
  baseURL,
  params,
  resolutions = DEFAULT_FIXED_RESOLUTIONS,
) =>
  resolutions
    .map(resolution => {
      const url = buildURL(baseURL, { ...params, dpr: resolution })
      return `${url} ${resolution}x`
    })
    .join(', ')

const buildFluidSrcSet = (baseURL, params, breakpoints) => {
  const { width } = params

  if (!breakpoints)
    breakpoints = DEFAULT_FLUID_BREAKPOINT_FACTORS.map(x => width * x)

  // Remove duplicates, sort by numberical value, and ensure maxWidth is added.
  const uniqSortedBreakpoints = [
    ...new Set([...breakpoints, width].sort((a, b) => a - b)),
  ]

  return uniqSortedBreakpoints
    .map(breakpoint => {
      const url = buildURL(baseURL, {
        ...params,
        width: breakpoint,
        height: undefined,
      })
      return `${url} ${breakpoint}w`
    })
    .join(', ')
}

const extractImageURLData = url => {
  const instance = new URL(url)
  const baseURL = instance.origin + instance.pathname
  const params = instance.searchParams

  return { baseURL, params }
}

export const getFixedGatsbyImage = (
  url,
  sourceWidth,
  sourceHeight,
  args = {},
) => {
  const { baseURL, params } = extractImageURLData(url)

  const rect = params.get('rect')
  const aspectRatio = sourceWidth / sourceHeight
  const width = args.width ?? DEFAULT_FIXED_WIDTH
  const height = args.height ?? Math.round(width / aspectRatio)
  const quality = args.quality

  const base64 = buildBase64URL(baseURL, { rect })
  const src = buildURL(baseURL, { width, height, rect, quality })
  const srcSet = buildFixedSrcSet(baseURL, { width, height, rect, quality })

  return {
    base64,
    aspectRatio,
    width,
    height,
    src,
    srcWebp: src,
    srcSet,
    srcSetWebp: srcSet,
  }
}

export const getFluidGatsbyImage = (
  url,
  sourceWidth,
  sourceHeight,
  args = {},
) => {
  const { baseURL, params } = extractImageURLData(url)

  const rect = params.get('rect')
  const aspectRatio = sourceWidth / sourceHeight
  const width = args.maxWidth ?? DEFAULT_FLUID_MAX_WIDTH
  const height = args.maxHeight ?? Math.round(width / aspectRatio)
  const quality = args.quality
  const srcSetBreakpoints = args.srcSetBreakpoints

  const base64 = buildBase64URL(baseURL, { rect })
  const src = buildURL(baseURL, { width, height, rect, quality })
  const srcSet = buildFluidSrcSet(
    baseURL,
    { width, height, rect, quality },
    srcSetBreakpoints,
  )

  return {
    base64,
    aspectRatio,
    src,
    srcWebp: src,
    srcSet,
    srcSetWebp: srcSet,
    // sizes,
  }
}
