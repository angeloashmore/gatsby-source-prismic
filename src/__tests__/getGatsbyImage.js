import { getFixedGatsbyImage, getFluidGatsbyImage } from '../getGatsbyImage'

const URL_SRC = 'https://images.prismic.io/reponame/imagehash.jpg'
const URL_SRC_WITH_RECT =
  'https://images.prismic.io/reponame/imagehash.jpg?rect=0,0,3600,1800'
const URL_HEIGHT = 1000
const URL_WIDTH = 2000

const safeParseInt = x =>
  x !== undefined && x !== null ? Number.parseInt(x) : x

const safeParseFloat = x =>
  x !== undefined && x !== null ? Number.parseFloat(x) : x

const extractURLData = url => {
  const instance = new URL(url)
  const params = instance.searchParams

  return {
    urlOrigin: instance.origin,
    urlPathname: instance.pathname,
    width: safeParseInt(params.get('w')),
    height: safeParseInt(params.get('h')),
    dpr: safeParseFloat(params.get('dpr')),
    rect: params.get('rect'),
    fit: params.get('fit'),
    quality: safeParseInt(params.get('q')),
    blur: safeParseInt(params.get('blur')),
    auto: params.get('auto'),
  }
}

const extractSrcSetURLData = srcSet => {
  const urls = srcSet.split(/,\s+/).map(x => x.split(' '))

  return urls.map(([url, descriptor]) => ({
    descriptor,
    url: extractURLData(url),
  }))
}

const extractImageData = ({ aspectRatio, base64, src, srcSet }) => ({
  aspectRatio,
  base64: extractURLData(base64),
  src: extractURLData(src),
  srcSet: extractSrcSetURLData(srcSet),
})

describe('getFixedGatsbyImage', () => {
  test('jpg without params', () => {
    const result = getFixedGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT)
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with rect without params', () => {
    const result = getFixedGatsbyImage(URL_SRC_WITH_RECT, URL_WIDTH, URL_HEIGHT)
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with width (600)', () => {
    const result = getFixedGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      width: 600,
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with height (400)', () => {
    const result = getFixedGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      height: 400,
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with width (600) and height (400)', () => {
    const result = getFixedGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      width: 600,
      height: 400,
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })
})

describe('getFluidGatsbyImage', () => {
  test('jpg without params', () => {
    const result = getFluidGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT)
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with rect without params', () => {
    const result = getFluidGatsbyImage(URL_SRC_WITH_RECT, URL_WIDTH, URL_HEIGHT)
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with max width (1000)', () => {
    const result = getFluidGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      maxWidth: 1000,
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with max height (600)', () => {
    const result = getFluidGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      maxHeight: 600,
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with max width (1000) and max height (600)', () => {
    const result = getFluidGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      maxWidth: 1000,
      maxHeight: 600,
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })

  test('jpg with srcSetBreakpoints', () => {
    const result = getFluidGatsbyImage(URL_SRC, URL_WIDTH, URL_HEIGHT, {
      srcSetBreakpoints: [400],
    })
    expect(extractImageData(result)).toMatchSnapshot()
    expect(result.srcSetWebp).toEqual(result.srcSet)
    expect(result.srcWebp).toEqual(result.src)
  })
})
