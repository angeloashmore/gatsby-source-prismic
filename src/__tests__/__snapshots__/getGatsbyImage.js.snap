// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFixedGatsbyImage jpg with width (600) 1`] = `
Object {
  "aspectRatio": 1.996007984031936,
  "base64": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=100&rect=0,0,3599,1803&fit=max&q=20&blur=15&auto=compress%2Cformat",
  "height": 301,
  "src": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "srcSet": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&dpr=1&q=50&auto=compress%2Cformat 1x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&dpr=1.5&q=50&auto=compress%2Cformat 1.5x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&dpr=2&q=50&auto=compress%2Cformat 2x",
  "srcSetWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&dpr=1&q=50&auto=compress%2Cformat 1x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&dpr=1.5&q=50&auto=compress%2Cformat 1.5x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&dpr=2&q=50&auto=compress%2Cformat 2x",
  "srcWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=600&h=301&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "width": 600,
}
`;

exports[`getFixedGatsbyImage jpg without params 1`] = `
Object {
  "aspectRatio": 1.996007984031936,
  "base64": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=100&rect=0,0,3599,1803&fit=max&q=20&blur=15&auto=compress%2Cformat",
  "height": 200,
  "src": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "srcSet": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&dpr=1&q=50&auto=compress%2Cformat 1x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&dpr=1.5&q=50&auto=compress%2Cformat 1.5x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&dpr=2&q=50&auto=compress%2Cformat 2x",
  "srcSetWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&dpr=1&q=50&auto=compress%2Cformat 1x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&dpr=1.5&q=50&auto=compress%2Cformat 1.5x, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&dpr=2&q=50&auto=compress%2Cformat 2x",
  "srcWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&h=200&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "width": 400,
}
`;

exports[`getFluidGatsbyImage jpg with max width (1000) 1`] = `
Object {
  "aspectRatio": 1.996007984031936,
  "base64": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=100&rect=0,0,3599,1803&fit=max&q=20&blur=15&auto=compress%2Cformat",
  "src": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&h=501&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "srcSet": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=250&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 250w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=500&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 500w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1000w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1500&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1500w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=2000&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 2000w",
  "srcSetWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=250&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 250w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=500&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 500w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1000w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1500&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1500w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=2000&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 2000w",
  "srcWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&h=501&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
}
`;

exports[`getFluidGatsbyImage jpg with max width (1000) and srcSetBreakpoints 1`] = `
Object {
  "aspectRatio": 1.996007984031936,
  "base64": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=100&rect=0,0,3599,1803&fit=max&q=20&blur=15&auto=compress%2Cformat",
  "src": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&h=501&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "srcSet": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 400w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1000w",
  "srcSetWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 400w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1000w",
  "srcWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1000&h=501&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
}
`;

exports[`getFluidGatsbyImage jpg without params 1`] = `
Object {
  "aspectRatio": 1.996007984031936,
  "base64": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=100&rect=0,0,3599,1803&fit=max&q=20&blur=15&auto=compress%2Cformat",
  "src": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=800&h=401&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
  "srcSet": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=200&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 200w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 400w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=800&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 800w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1200&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1200w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1600&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1600w",
  "srcSetWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=200&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 200w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=400&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 400w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=800&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 800w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1200&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1200w, https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=1600&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat 1600w",
  "srcWebp": "https://images.prismic.io/pvadev/7802e8de-6b1a-4e65-b6bc-69cde942f727_LG_6250R_crop.JPG?w=800&h=401&rect=0,0,3599,1803&fit=max&q=50&auto=compress%2Cformat",
}
`;
